### PR TITLE
fix: detailed error for store init failure and deferred execution of closers

### LINKED ISSUE #1684: program panic when failed to initialize etcd store is unreasonable
## Issue description

Say my etcd cluster contains some corrupted data, which causes the manager api launching failed, the manager-api just crashes and the output is not so friendly.

## Expected behavior

Output some messages to hint at this failure and exit.

## Screenshots

image

## Environment

* apisix version (cmd: `apisix version`):
* OS (cmd: `uname -a`):
* OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`):
* etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API): 3.5.0-alpha0
* apisix-dashboard version, if have: 2.5
* Browser version, if have:

### PR DESCRIPTION
**Pull request by bisakhmondal**
Please answer these questions before submitting a pull request, **or your PR will get closed**.

**Why submit this pull request?**

- [x] Bugfix
- [ ] New feature provided
- [x] Improve performance
- [ ] Backport patches

**What changes will this PR take into?**

1. Closing of all opened resource has been put into deferred execution. So incase any panic occurs during runtime, the respective closers of the resource gets executed safely. Till now, for panics, the main goroutine was getting closed ungracefully.

2. Detailed error description if etcd contains corrupted data and exit with nonzero error code.
image

**Related issues**

resolves #1684 

**Checklist:**

- [x] Did you explain what problem does this PR solve? Or what new features have been added?
- [ ] Have you added corresponding test cases?
- [ ] Have you modified the corresponding document?
- [x] Is this PR backward compatible? If it is not backward compatible, please discuss on the mailing list first

### CODE REVIEW

**Comment by bisakhmondal** [comment]
docker registry seems temporary down. All requests are failing.

**Comment by nic-chen** [comment]
@bisakhmondal  I think we should add a cli test case for it

File: api/cmd/managerapi.go [hunk 1]
```diff
@@ -99,7 +108,9 @@ func NewManagerAPICommand() *cobra.Command {
 			}
 			if err := store.InitStores(); err != nil {
 				log.Errorf("init stores fail: %w", err)
-				panic(err)
+				fmt.Printf("%s\n", err)
```
Comment (tokers) [inline api/cmd/managerapi.go:111]: We should output exceptions to stderr, not stdout.

File: api/internal/core/store/store.go [hunk 2]
```diff
@@ -105,6 +105,7 @@ func (s *GenericStore) Init() error {
 		key := ret[i].Key[len(s.opt.BasePath)+1:]
 		objPtr, err := s.StringToObjPtr(ret[i].Value, key)
 		if err != nil {
+			fmt.Println("Error occurred while initializing logical store: ", s.opt.BasePath)
```
Comment (tokers) [inline api/internal/core/store/store.go:108]: Ditto.

File: api/cmd/managerapi.go (hunk 1 above)
Comment (bisakhmondal) [reply api/cmd/managerapi.go:111]: Oops!😅😂

**Comment by bisakhmondal** [comment]
> @bisakhmondal I think we should add a cli test case for it

Okay😁. Adding...

**Comment by bisakhmondal** [comment]
Hii @nic-chen, could you please have a look here when you have some time
https://github.com/apache/apisix-dashboard/pull/1705/checks?check_run_id=2272703281#step:5:720

File: api/test/shell/cli_test.sh [hunk 3]
```diff
@@ -457,6 +457,23 @@ fi
 ./manager-api stop
 sleep 6
 
+# test manager-api output for bad data on etcd
+# make a dummy entry
+./etcd-v3.4.14-linux-amd64/etcdctl put /apisix/routes/unique1 "{\"id\":}"
```
Comment (nic-chen) [inline api/test/shell/cli_test.sh:462]: I think you could try to sleep 1 after put to etcd

Comment (bisakhmondal) [reply api/test/shell/cli_test.sh:462]: Still not working😶

**Review by starsz — COMMENTED** [review]
LGTM, but CI is failed.

**Comment by bisakhmondal** [comment]
> LGTM, but CI is failed.

image

Yep. Ain't getting why😄!!
Why is manager-api starting at all😅?

**Review by tokers — APPROVED** [review]

**Comment by nic-chen** [comment]
hi @bisakhmondal 
Have you tested it locally?

**Comment by bisakhmondal** [comment]
> hi @bisakhmondal
> Have you tested it locally?

Hii @nic-chen , yes many times inside an isolated docker container (ubuntu image) just like CI. There is literally no issue. Everything is working fine.
**Snippet used**:
```shell
# test manager-api output for bad data on etcd
# make a dummy entry
etcdctl put /apisix/routes/unique1 "{\"id\":}"
sleep 2
## for debugging purpose ------------
#./etcd-v3.4.14-linux-amd64/etcdctl get --keys-only --prefix /apisix/routes
#./etcd-v3.4.14-linux-amd64/etcdctl get /apisix/routes/unique1
##-------------------------------

./manager-api 2>man-api.err &
sleep 4

if [[ `cat man-api.err | grep -c "Error occurred while initializing logical store:  /apisix/routes"` -ne '1' ||
`cat man-api.err | grep -c "json unmarshal failed"` -ne '1' ]];then
  echo "manager api failed to stream error on stderr for bad data"
  exit 1
fi
echo "Test Successful"
# delete dummy entry
etcdctl del /apisix/routes/unique1
# just to make sure
./manager-api stop
```

**Output**:
You can see there is no previous entry for `/apisix/routes/unique1` and no file name `man-api.err` as well (just to make sure) where the stderr stream is getting flushed.
image

**Comment by nic-chen** [comment]
hi @bisakhmondal  I think you could write some log to debug on CI.

**Comment by bisakhmondal** [comment]
> hi @bisakhmondal I think you could write some log to debug on CI.

Okay. Sounds good.

**Comment by nic-chen** [comment]
hi @bisakhmondal 
any progress?

**Comment by bisakhmondal** [comment]
> hi @bisakhmondal
> any progress?

Hii. Got caught up in a few work. Going to test it tonight :)

File: api/cmd/managerapi.go [hunk 4]
```diff
@@ -93,13 +102,16 @@ func NewManagerAPICommand() *cobra.Command {
 				return newMws
 			}
 
+			fmt.Printf("DEBUG-ETCDCONFIG: %+v\n\n", *conf.ETCDConfig)
```
Comment (nic-chen) [inline api/cmd/managerapi.go:105]: please remove debug code

Comment (bisakhmondal) [reply api/cmd/managerapi.go:105]: yes yes. Just 1 min.
image
Actually, it was the root cause. The endpoint was different, haha.

**Review by nic-chen — APPROVED** [review]

**Review by starsz — APPROVED** [review]

**Comment by juzhiyuan** [comment]
image

@nic-chen Looks like CI pending again?

**Comment by nic-chen** [comment]
same as #1781

**Comment by bisakhmondal** [comment]
Still the same!! The CI is stalled.

**Comment by juzhiyuan** [comment]
emmm what happened then?

**Comment by bisakhmondal** [comment]
I guess the PR is complete. Anything left that needs to be done? @nic-chen

**Comment by nic-chen** [comment]
@bisakhmondal  
CI still pending.. 
Would you try to rebase? 
Or we can only use the last move, open a new branch, cherry-pick your changes to the new branch, and then submit a new PR.

**Comment by bisakhmondal** [comment]
Cool. Let me try rebasing first. If still doesn't work, then going with the second option.

**Comment by bisakhmondal** [comment]
image
Nope, it didn't work

**Comment by bisakhmondal** [comment]
Even reopening doesn;t work :(

**Pull request closed without merging**